src/apps/monero/protocol/tsx_sign_builder.py

```diff
@@ -0,0 +1,1415 @@
+
+        # Rsig data
+        self.rsig_type = tsx_data.rsig_data.rsig_type
+        self.rsig_grp = tsx_data.rsig_data.grouping
+        self.rsig_offload = self.rsig_type > 0 and self.output_count > 2
+        self.use_bulletproof = self.rsig_type > 0
+        self.use_simple_rct = self.input_count > 1 or self.rsig_type != 0
+
+        # Provided tx key, used mostly in multisig.
+        if len(tsx_data.use_tx_keys) > 0:
+            for ckey in tsx_data.use_tx_keys:
+                crypto.check_sc(crypto.decodeint(ckey))
+
+            self.gen_r(use_r=crypto.decodeint(tsx_data.use_tx_keys[0]))
+            self.additional_tx_private_keys = [
+                crypto.decodeint(x) for x in tsx_data.use_tx_keys[1:]
+            ]
+
+        # Additional keys w.r.t. subaddress destinations
+        class_res = classify_subaddresses(tsx_data.outputs, self.change_address())
+        num_stdaddresses, num_subaddresses, single_dest_subaddress = class_res
+
+        # if this is a single-destination transfer to a subaddress, we set the tx pubkey to R=r*D
+        if num_stdaddresses == 0 and num_subaddresses == 1:
+            self.r_pub = crypto.scalarmult(
+                crypto.decodepoint(single_dest_subaddress.spend_public_key), self.r
+            )
+
+        self.need_additional_txkeys = num_subaddresses > 0 and (
+            num_stdaddresses > 0 or num_subaddresses > 1
+        )
+        self._mem_trace(4, True)
+
+        # Extra processing, payment id
+        self.tx.version = 2
+        self.tx.unlock_time = tsx_data.unlock_time
+        self.process_payment_id(tsx_data)
+        await self.compute_sec_keys(tsx_data)
+        gc.collect()
+
+        # Iterative tx_prefix_hash hash computation
+        self._tprefix_update()
+        gc.collect()
+
+        # Final message hasher
+        self.full_message_hasher.init(self.use_simple_rct)
+        self.full_message_hasher.set_type_fee(self.get_rct_type(), self.get_fee())
+
+        # Sub address precomputation
+        if tsx_data.account is not None and tsx_data.minor_indices:
+            self.precompute_subaddr(tsx_data.account, tsx_data.minor_indices)
+        self._mem_trace(5, True)
+
+        # HMAC outputs - pinning
+        hmacs = []
+        for idx in range(self.num_dests()):
+            c_hmac = await self.gen_hmac_tsxdest(tsx_data.outputs[idx], idx)
+            hmacs.append(c_hmac)
+            gc.collect()
+
+        self._mem_trace(6)
+
+        from trezor.messages.MoneroTransactionInitAck import MoneroTransactionInitAck
+        from trezor.messages.MoneroTransactionRsigData import MoneroTransactionRsigData
+
+        rsig_data = MoneroTransactionRsigData(offload_type=self.rsig_offload)
+        return MoneroTransactionInitAck(
+            in_memory=False,
+            many_inputs=True,
+            many_outputs=True,
+            hmacs=hmacs,
+            rsig_data=rsig_data,
```

Comment:
Mmm, so we're returning `MoneroTransactionInitAck` with `MoneroTransactionRsigData` just to communicate `self.rsig_offload`? Also `in_memory=False, many_inputs=True, many_outputs=True` is hardcoded but might not be in the future?

Reply:
I decided to group all rsig-related data to the `MoneroTransactionRsigData` as there are more possible offloading techniques. In this case, I found it acceptable as the message has only one field filled in so I found the overhead negligible.

`in_memory=False, many_inputs=True, many_outputs=True` are like 2 days old optimizations as we discussed to simplify the protocol and strip down the optimizations in order to unify the flow. 

In general, I am hesitant to modify messages structures too often because it usually requires PR and merging which is needed so I can continue with the work.

We can mark `in_memory=False, many_inputs=True, many_outputs=True` as deprecated and remove it in the next PR. What do you propose to do with `MoneroTransactionRsigData`?

Reply:
> We can mark in_memory=False, many_inputs=True, many_outputs=True as deprecated and remove it in the next PR. 

Sounds good. We can make an issue if you want to track those things.

> What do you propose to do with MoneroTransactionRsigData?

Let's keep it for now. It just seemed a bit weird to nest a MoneroTransactionRsigData in a MoneroTransactionInitAck, which does not contain any Rsig data obvisouly, because it's a first message. But it's a nit-pick, let's keep it for now.